terracotta/drivers/base_classes.py

```diff
@@ -0,0 +1,135 @@
+"""
+
+import contextlib
+import functools
+from abc import ABC, abstractmethod
+from collections import OrderedDict
+from typing import (Any, Callable, Dict, List, Mapping, Optional, Sequence,
+                    Tuple, TypeVar, Union)
+
+KeysType = Mapping[str, str]
+MultiValueKeysType = Mapping[str, Union[str, List[str]]]
+Number = TypeVar('Number', int, float)
+T = TypeVar('T')
+
+
+def requires_connection(
+    fun: Callable[..., T] = None, *,
+    verify: bool = True
+) -> Union[Callable[..., T], functools.partial]:
+    if fun is None:
+        return functools.partial(requires_connection, verify=verify)
+
+    @functools.wraps(fun)
+    def inner(self: MetaStore, *args: Any, **kwargs: Any) -> T:
+        assert fun is not None
+        with self.connect(verify=verify):
+            return fun(self, *args, **kwargs)
+
+    return inner
+
+
+class MetaStore(ABC):
+    """Abstract base class for all Terracotta metadata backends.
+
+    Defines a common interface for all metadata backends.
+    """
+    _RESERVED_KEYS = ('limit', 'page')
+
+    @property
+    @abstractmethod
+    def db_version(self) -> str:
+        """Terracotta version used to create the database."""
+        pass
+
+    @property
+    @abstractmethod
+    def key_names(self) -> Tuple[str, ...]:
+        """Names of all keys defined by the database."""
+        pass
+
+    @abstractmethod
+    def __init__(self, url_or_path: str) -> None:
+        self.path = url_or_path
+
+    @classmethod
+    def _normalize_path(cls, path: str) -> str:
+        """Convert given path to normalized version (that can be used for caching)"""
+        return path
+
+    @abstractmethod
+    def create(self, keys: Sequence[str], *,
+               key_descriptions: Mapping[str, str] = None) -> None:
+        """Create a new, empty database"""
+        pass
+
+    @abstractmethod
+    def connect(self, verify: bool = True) -> contextlib.AbstractContextManager:
+        """Context manager to connect to a given database and clean up on exit.
+
+        This allows you to pool interactions with the database to prevent possibly
+        expensive reconnects, or to roll back several interactions if one of them fails.
+        """
+        pass
+
+    @abstractmethod
+    def get_keys(self) -> OrderedDict:
+        """Get all known keys and their fulltext descriptions."""
+        pass
+
+    @abstractmethod
+    def get_datasets(self, where: MultiValueKeysType = None,
+                     page: int = 0, limit: int = None) -> Dict[Tuple[str, ...], Any]:
+        """Get all known dataset key combinations matching the given constraints,
+        and a path to retrieve the data
+        """
+        pass
+
+    @abstractmethod
+    def get_metadata(self, keys: KeysType) -> Optional[Dict[str, Any]]:
+        """Return all stored metadata for given keys."""
+        pass
+
+    @abstractmethod
+    def insert(self, keys: KeysType,
+               path: Any, **kwargs: Any) -> None:
```

Comment:
Is there a reason to have the `**kwargs` here? If not, then let's remove it

Reply:
I guess it's to allow for backend-specific parameters. But actually, `TerracottaDriver` calls `MetaStore.insert()` like this
```python
self.meta_store.insert(
            keys=keys,
            path=override_path or path,
            metadata=metadata,
            **kwargs
        )
```
So in order for the abstract interface to conform to that, it should look like this
```python
@abstractmethod
def insert(self, keys: KeysType, path: Any, 
           metadata: Optional[Mapping[str, Any]], **kwargs: Any) -> None:
```

Reply:
> I used this to indicate that derived classes may take additional keyword args.

Right, I see. That makes sense. But should we specify 
```python
    def insert(self, keys: KeysType, path: str, metadata: Mapping[str, Any] = None) -> None:
```
instead? Unless we need that extra flexibility, I would prefer to specify arguments exactly.

Note that I have actually quietly done that in `RelationalMetaStore`, which does not accept arbitrary `**kwargs` in its `insert` (and wouldn't do anything with them, in any case) -- actually making `Terracotta` driver fail if you give additional arguments to its `insert` when using a `RelationalMetaStore`. I've done the same for `compute_metadata` in  `RasterStore` and `TerracottaDriver`.

Reply:
> So in order for the abstract interface to conform to that, it should look like this

Basically, I'm advocating that we shall do this the other way around, unless we actually need that flexibility. If possible, which I believe it is here, I think it's better to design the interfaces to describe our usage for them, rather than making them take arbitrary `**kwargs`, which subsequently requires us to be less specific down the line.

Reply:
Fine with me.

Reply:
Can we image `insert`ing things besides these
```python
def insert(
        self,
        keys: KeysType,
        path: str, *,
        metadata: Mapping[str, Any] = None
    ) -> None
```
?
If so, this will require a slight modification to `RelationalMetaStore.insert`. If not, then I prefer to specify the arguments.

Reply:
As long as all driver combinations have the same signature I agree that we should be explicit about the arguments.

Reply:
All relational drivers does, at least 😄 I'll specify it. We can always revert if we ever encounter a situation where we want to insert other things

Reply:
Can I specify `path` from `Any` to `str` as well?

Reply:
Yes